### ISSUE TITLE
utils._ensure_real_directory() very slow in some cases

### ISSUE THREAD

**Issue by Cynical-Optimist** [buildstream-migration/bst-staging#174]
[See original issue on GitLab](https://gitlab.com/BuildStream/buildstream/-/issues/174)
In GitLab by [[Gitlab user @samthursfield]](https://gitlab.com/samthursfield) on Dec 19, 2017, 18:10

I have been experiencing slow staging performance. As a data point, `bst shell base.bst` inside the gnome-modulesets.git project takes nearly 2 minutes to actually open a shell prompt.

I profiled the operation with Pyflame and discovered a lot of time is spent during staging in `utils.link_files()` (26 seconds) > `utils._ensure_real_directory()` > `os.path.realpath()` (~19 seconds).

Looking at the [code for os.path.realpath()](https://github.com/python/cpython/blob/master/Lib/posixpath.py#L382), it seems like the issue is just that the Python standard library implementation of that function is very inefficient.

Removing the `os.path.realpath()` calls from `utils._ensure_real_directory()` saves around 20 seconds of runtime; we shouldn't remove this check, but we should definitely look for how to speed it up. One option would be to try importing the `realpath()` from the system's C library using the ctypes module and calling that instead.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @samthursfield]](https://gitlab.com/samthursfield) on Dec 19, 2017, 18:18

![shell-profile.svg](/uploads/35497e1f279f2c1abac65df661b75a1a/shell-profile.svg)

This is the profile. It doesn't follow subprocesses, so a lot of the time is just shown as 'idle'.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Dec 19, 2017, 18:37

Looks like we need more caching I think:
* If we walk the filesystem only once and cache the result, this should help things
* For calls to `os.realpath()`, I dont see why we would ever have to do this more than once for a given file in a given artifact
* This will require some form of IPC, or have the child process write out a manifest
* This is the idea behind #82, whether we actually store the manifest in an artifact or not; discovering the manifest should ideally be a one-off activity

I *think* the right solution should be along these lines:
* Add some prepare() method to Element and ArtifactCache classes
  * ArtifactCache.prepare() - Does any one-shot work for the artifact (means preparing the extract directory), this is a noop if it's already done.
  * Element._prepare() calls into the ActifactCache.prepare(), and as a result; loads it's full manifest
* Redirect filesystem walks
  * Element staging APIs should no longer do the filesys walks, but perform the split rule globbing directly on loaded/cached file lists
* Add PrepareQueue to build plans
  * PrepareQueue is always placed after the BuildQueue (an element can only be *"prepared"* after it's been *"cached"*).
  * PrepareQueue's element ready condition is that the element to be prepared is already *"cached"*.
  * PrepareQueue can only ever skip an element which has no reverse dependencies in the pipeline (or just never skip anything).
  * BuildQueue's element ready condition requires that build dependencies be *"prepared"*

Note this is still entirely orthogonal to the decision in #82, if we store manifests and load them later; it should be in the prepare stage.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @samthursfield]](https://gitlab.com/samthursfield) on Dec 20, 2017, 11:20

changed title from **Python os.realpath() function is too slow** to **Python os.{+path.+}realpath() function is too slow**

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @samthursfield]](https://gitlab.com/samthursfield) on Dec 20, 2017, 11:20

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @samthursfield]](https://gitlab.com/samthursfield) on Dec 20, 2017, 11:30

Caching of manifests should help other performance issues, but the problem here occurs even with something like `bst shell` where we only look at each artifact once.

I traced the calls we make to `os.path.realpath()` during a single `bst shell` operation and found that we call it for the same directories over and over again, probably once per file contained within each directory. Hopefully it's not too hard to be a bit smarter there.

I also realised that on this machine the cache has a symlink in its path, which might explain why I see this issue worse than everyone else.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @samthursfield]](https://gitlab.com/samthursfield) on Dec 20, 2017, 12:17

changed title from **{-Python os.path.realpath() function is too slow-}** to **{+utils._ensure_real_directory() very slow in some cases+}**

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @samthursfield]](https://gitlab.com/samthursfield) on Dec 20, 2017, 17:17

mentioned in merge request !198

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Dec 20, 2017, 19:38

> the problem here occurs even with something like bst shell where we only look at each artifact once.

Fair enough.

Is it also probable that we are calling `os.path.realpath()` more than once on the same directory ?

Maybe we can still speed this up a little with some caching ?

I suppose that `os.path.realpath()` is loopy, which also leads me to believe there should be a more optimized way of still discovering files real paths but with a less complex loop.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @samthursfield]](https://gitlab.com/samthursfield) on Dec 22, 2017, 12:20

mentioned in merge request !201

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on Jan 11, 2018, 16:32

assigned to [[Gitlab user @jonathanmaw]](https://gitlab.com/jonathanmaw)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @samthursfield]](https://gitlab.com/samthursfield) on Jan 19, 2018, 14:24

mentioned in commit dc8b923d0afd6cd854edad8759ba49819fe3dcd1

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @samthursfield]](https://gitlab.com/samthursfield) on Jan 19, 2018, 14:28

mentioned in merge request !246

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @samthursfield]](https://gitlab.com/samthursfield) on Jan 19, 2018, 15:19

mentioned in commit 33c0c7ab2f6eaed0bb5cd6d7993e80199605837a

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @samthursfield]](https://gitlab.com/samthursfield) on Jan 19, 2018, 15:19

mentioned in merge request !247

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @samthursfield]](https://gitlab.com/samthursfield) on Feb 12, 2018, 10:26

mentioned in commit 0f430622e4539b5cc9209384e32980dc086c29df

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Feb 12, 2018, 10:54

closed via merge request !247
